### PR TITLE
refactor: querier<>ingester flight protocol adjustments

### PR DESCRIPTION
**Pull request by crepererum**
This makes a few adjustments to the querier<>ingester flight protocol.

Query Scope
===========
The querier will request data for ALL sequencer IDs for now. There is
no reason to have a request per sequencer ID. We can add a range/set
filter later if we want, but this is not required for now.

Partition-level
===============
The only time when the querier cares about sequencer IDs (i.e. sharding)
at all is when it selects which ingesters to ask for unpersisted data
(this is currently not implemented, it just asks all ingesters).
Afterwards the querier only cares about partitions (which are bound to
specific sequencers anyways) because this is the level where parquet
file persistence and compaction as well as deduplication happen. So we
make partitions a first-class citizen in the ingester response.

Metadata VS RecordBatches
=========================
The global app-metadata will list all partitions and their max
persisted parquet files and tombstones (theoretically tombstones are at
table-level, but the ingester could in the future break them down to the
partition-level). Then it receives a stream of record batches. Each
record batch is tagged (via key-value metadata in its schema) so it can
be assigned to a partition. At the moment the ingester returns 0 or 1
batches per unpersisted partition (0 in case we've filtered out all the
data via the predicate), but in the future it is free to return multiple
batches. This setup gives the ingester more freedom over memory
management and (potentially parallel) query processing, while at the
same time keeps the set of duplicated information minimal and allows
easy extensions (since the global metadata is a full-blown protobuf
message).

Querier
=======
At the moment the querier ignores all the metdata. Follow-up PRs will
change that.

### CODE REVIEW

File: generated_types/protos/influxdata/iox/ingester/v1/query.proto [hunk 1]
```diff
@@ -43,11 +44,30 @@ message IngesterQueryResponseMetadata {
...
-  optional int64 parquet_max_sequence_number = 3;
+  // Was max sequence number persisted for this table
+  reserved "parquet_max_sequence_number";
+  reserved 3;
+
+  // Was max sequence number for a tombstone associated with this table
+  reserved "tombstone_max_sequence_number";
+  reserved 4;
+
+  // Contains status for every partition that has unpersisted data.
```
Comment (alamb) [inline generated_types/protos/influxdata/iox/ingester/v1/query.proto:55]:
```suggestion
  // Map partition_id to status for every partition that has unpersisted data.
```

File: generated_types/protos/influxdata/iox/ingester/v1/query.proto [hunk 2]
```diff
@@ -43,11 +44,30 @@ message IngesterQueryResponseMetadata {
...
+  //
+  // If a partition does NOT appear within this map, then either all data was persisted or the ingester has never seen
+  // data for this partition. In either case the querier may just read all parquet files for the missing partition.
```
Comment (alamb) [inline generated_types/protos/influxdata/iox/ingester/v1/query.proto:58]:
```suggestion
  // data for this partition. In either case the querier may just read all parquet files for the missing partition
  // and ensure it has a complete dataset
```

File: influxdb_iox/tests/end_to_end_ng_cases/ingester.rs [hunk 3]
```diff
@@ -49,10 +50,18 @@ async fn ingester_flight_api() {
         .await
         .unwrap();
 
-    assert!(performed_query
-        .app_metadata()
-        .parquet_max_sequence_number
-        .is_none());
+    let unpersisted_partitions = &performed_query.app_metadata().unpersisted_partitions;
```
Comment (alamb) [inline influxdb_iox/tests/end_to_end_ng_cases/ingester.rs:53]: is this point of this test to validate that the `unpersisted_partition` field is actually returned?

File: ingester/src/data.rs [hunk 4]
```diff
@@ -880,6 +876,14 @@ impl TableData {
     }
 }
 
+#[derive(Debug)]
+pub(crate) struct InFlightData {
```
Comment (alamb) [inline ingester/src/data.rs:880]: I suggest calling this struct something like `UnpersistedData` or `UnpersistedPartitionData` to be more consistent with the terminology used in the rest of the ingester. I worry that using `Flight` might get confusing with `Arrow Flight`

Comment (alamb) [reply ingester/src/data.rs:880]: I also suggest adding a docstring explaining what the struct is used for, like

```rust
/// Read only copy of the unpersisted data for a partition
/// in the ingester

File: ingester/src/data.rs [hunk 5]
```diff
@@ -880,6 +876,14 @@ impl TableData {
...
+    pub partition_id: PartitionId,
+    pub non_persisted: Vec<Arc<SnapshotBatch>>,
+    pub persisting: Option<QueryableBatch>,
+    pub parquet_max_sequence_number: Option<SequenceNumber>,
```
Comment (alamb) [inline ingester/src/data.rs:884]: Perhaps should this be `partition_status: PartitionStatus`? It seems strange to only have the parquet file max sequence here

File: ingester/src/data.rs [hunk 6]
```diff
@@ -1378,6 +1382,19 @@ pub struct QueryableBatch {
...
+pub struct PartitionStatus {
+    /// Max sequence number persisted
+    pub parquet_max_sequence_number: Option<SequenceNumber>,
+
+    /// Max sequence number for a tombstone
+    pub tombstone_max_sequence_number: Option<SequenceNumber>,
+}
+
 /// Response sending to the query service per its request defined in IngesterQueryRequest
 pub struct IngesterQueryResponse {
```
Comment (alamb) [inline ingester/src/data.rs:1399]: The assymetry of having `IngesterQueryRequest` in `data_types2` but `IngesterQueryRsponse` only in the `ingester` is strange to me. 

I know you didn't make it like this in this PR I just noticed it while reading

File: querier/src/ingester/mod.rs [hunk 7]
```diff
@@ -136,9 +136,6 @@ impl IngesterConnection for IngesterConnectionImpl {
 
             let ingester_query_request = IngesterQueryRequest {
                 namespace: namespace_name.to_string(),
-                // TODO: change protocol to search all sequencers...
```
Comment (alamb) [inline querier/src/ingester/mod.rs:139]: 👌

File: ingester/src/querier_handler.rs [hunk 8]
```diff
@@ -49,10 +58,11 @@ pub enum Error {
     CollectStream { source: DataFusionError },
 
     #[snafu(display(
-        "No Table Data found for the given sequencer id {}, namespace name {}, table name {}", sequencer_id.get(), namespace_name, table_name
+        "No Table Data found for the given namespace name {}, table name {}",
```
Comment (alamb) [inline ingester/src/querier_handler.rs:61]: this is the error you were mentioning this morning, right? Specifically the querier should ignore such errors

File: ingester/src/querier_handler.rs [hunk 9]
```diff
@@ -72,31 +85,92 @@ pub async fn prepare_data_to_querier(
...
+            Some(table_data) => table_data,
+            None => {
+                continue;
+            }
+        };
+
+        let (tombstone_max_sequence_number, inflight_data) = {
+            let table_data = table_data.read().await;
+            (
+                table_data.tombstone_max_sequence_number(),
```
Comment (alamb) [inline ingester/src/querier_handler.rs:109]: I think if you put the tombstone into `inflight_data()` this code would be even more concise

File: ingester/src/querier_handler.rs [hunk 10]
```diff
@@ -72,31 +85,92 @@ pub async fn prepare_data_to_querier(
...
+                continue;
+            }
+        };
+
+        let table_data = match namespace_data.table_data(&request.table) {
+            Some(table_data) => table_data,
+            None => {
+                continue;
+            }
+        };
```
Comment (alamb) [inline ingester/src/querier_handler.rs:104]: I wonder if writing this code like would be nicer

```suggestion
        let maybe_table_data = sequencer_data.namespace(&request.namespace)
          .map(|namespace_data| namespace_data.table_data(&request.table);
     
          match maybe_table_data {
            Some(table_data) => table_data,
            None => {
                continue;
            }
        };     
```

File: ingester/src/querier_handler.rs (hunk 8 above)
Comment (crepererum) [reply ingester/src/querier_handler.rs:61]: yes

File: ingester/src/data.rs (hunk 5 above)
Comment (crepererum) [reply ingester/src/data.rs]: done

File: ingester/src/querier_handler.rs (hunk 9 above)
Comment (crepererum) [reply ingester/src/querier_handler.rs]: done

File: ingester/src/data.rs [hunk 11]
```diff
@@ -880,6 +879,15 @@ impl TableData {
     }
 }
 
+/// Read only copy of the unpersisted data for a partition in the ingester for a specific partition.
```
Comment (alamb) [inline ingester/src/data.rs:882]: 😍

**Review by alamb — APPROVED** [review]

File: ingester/src/querier_handler.rs [hunk 12]
```diff
@@ -210,11 +289,15 @@ pub async fn query(
...
-        ),
+        Selection::Some(columns) => {
+            let schema = data.schema();
+            Some(
+                columns
+                    .iter()
+                    .flat_map(|&column_name| schema.find_index_of(column_name))
+                    .collect(),
+            )
+        }
```
Comment (NGA-TRAN) [inline ingester/src/querier_handler.rs:300]: Is this code more optimal than the one above?

**Review by NGA-TRAN — APPROVED** [review]
LGTM

File: ingester/src/querier_handler.rs (hunk 12 above)
Comment (crepererum) [reply ingester/src/querier_handler.rs:300]: No, but it turns out that `compute_select_indices` errors when a column was not found which should be fine (I guess?)

**Pull request merged**